src/ray/core_worker/profiling.h

```diff
@@ -33,14 +34,14 @@ class Profiler {
   // until they are flushed.
   rpc::ProfileTableData rpc_profile_data_ GUARDED_BY(mu_);
 
-  gcs::RedisGcsClient &gcs_client_;
+  const std::shared_ptr<gcs::RedisGcsClient> gcs_client_;
```

Comment:
Oh sorry, I didn't intend to mean the class member should be `const`, just the parameters should be `const &`. Class members in C++ should pretty much always be non-`const`, non-references.

Reply:
I hadn't made any changes yet. I understand not having reference members, but why not `const`? That seems to add some semantic information if nothing else - "this pointer will only be modified during construction and destruction."

Reply:
It's a bit counterintuitive at first, and it's extremely tempting if you're coming from e.g. a C#/Java standpoint where classes have reference semantics, but when you try it after a while, you see it doesn't actually buy you anything tangible in C++, and has an annoying habit of getting in your way over and over again. Conceptually, it's because it just doesn't really make sense in C++'s value semantics (unlike, say, Java's) to have some members be `const` and some not; it pretty much never makes sense to assign some members and not others.

If you want some practical reasons, some issues I can think of off the top of my head include the following (there are quite possibly more than I can't think of)

- It suppresses moves (because you can't move _from_ these members), which will make any code that attempts to copy the class inefficient for no good reason, since they now have to copy.

- It suppresses assignments, which makes it difficult to use the class in a container (since now you can't assign to each member in the container).

- If you really want to suppress moves and assignments, you should actually delete those members explicitly to document that fact, not do it implicitly like this.

- It triggers various compiler warnings (e.g. "assignment operator has been deleted") that could otherwise be occasionally helpful, but which now get incredibly noisy and which you'll want to turn off.

A lot of these reasons apply to references in a similar way as well, which is why I mentioned both together.